### PR TITLE
Keep loss_scale and Whole Loss Subgraph in FP32 during Mixed Precision Training

### PR DESCRIPTION
**Pull request by iK1D**
Refactor the mixed precision transformer to keep loss_scale input and the whole loss subgraph in FP32 mode during mixed precision training.

**Motivation and Context**
- Current mixed precision transformer only converts loss and loss grad Ops to FP32, the subgraph after loss and the loss output is in FP16 and at the end times FP16 loss_scale to start BP. This is different from the official mixed precision training document that requires loss_scale in FP32 type and times to a FP32 loss value.
- PyTorch's BERT and UniLM v2 implementations keep the whole loss subgraph (starts from loss Ops and ends with loss grad Ops) in FP32.
- Ideal way is exporter can annotate and tell which nodes need to be kept in FP32, but since we can't get such information for now, we use loss and loss grad Ops to determine the loss subgraph.
- With this change, no perf regression observed for current BERT and UniLM v2 model.
- Will update some correctness test result.

### CODE REVIEW

File: orttraining/orttraining/core/graph/mixed_precision_transformer.cc [hunk 1]
```diff
@@ -70,10 +66,33 @@ bool IsFP32(const std::unordered_map<std::string, std::vector<int>>& map, std::s
   }
 }
 
+static const std::string Loss_Scale_Input = "loss_scale";
+static const std::string Scaled_Loss_Node = "scaled_loss";
+static const std::string Scaled_Loss_Grad_Mul_Node = "scaled_loss_Grad/Mul_0";
+static const std::string Scaled_Loss_Grad_ReduceSum_Node = "scaled_loss_Grad/ReduceSum_1";
+
```
Comment (weixingzhang) [inline orttraining/orttraining/core/graph/mixed_precision_transformer.cc:73]: This is not reliable way to check loss subgraph by checking the names of inputs/outputs. Right?

File: orttraining/orttraining/core/graph/mixed_precision_transformer.cc [hunk 2]
```diff
@@ -49,15 +47,13 @@ static const std::unordered_map<std::string, std::vector<int>> stage1_fp32_node_
...
+// It's possibile we will have more FP32 nodes added, this map will also be extended.
 static const std::unordered_map<std::string, std::vector<int>> stage2_fp32_node_args = {
     {"TrainableDropout", {1}},
     {"TrainableDropoutGrad", {2}},
     {"Dropout", {1}},
     {"DropoutGrad", {2}},
-    {"SparseSoftmaxCrossEntropy", {0, 2}},
-    {"SparseSoftmaxCrossEntropyGrad", {0, 1, 3}},
-    {"SoftmaxCrossEntropyLoss", {0, 2}},
-    {"SoftmaxCrossEntropyLossGrad", {0, 1, 3}},
```
Comment (weixingzhang) [inline orttraining/orttraining/core/graph/mixed_precision_transformer.cc:60]: Surprised to see loss scale in FP16. It is definitely a bug in the implementation. What does loss subgraph for BERT look like? If loss scale is in FP32, then the output of SparseSoftmaxCrossEntropy should not be casted to FP16.

File: orttraining/orttraining/core/graph/mixed_precision_transformer.cc (hunk 1 above)
Comment (iK1D) [reply orttraining/orttraining/core/graph/mixed_precision_transformer.cc:73]: You are right. Maybe I can try to get these node from loss_scale input by calling GetConsumers from graph.

File: orttraining/orttraining/core/graph/mixed_precision_transformer.cc (hunk 2 above)
Comment (iK1D) [reply orttraining/orttraining/core/graph/mixed_precision_transformer.cc:60]: Current logic just add CastF32 right before and CastF16 right after the two SparseSoftmaxCrossEntropy and SparseSoftmaxCrossEntropyGrad node only. All other parts of the graph are in FP16, including those nodes between SparseSoftmaxCrossEntropy and SparseSoftmaxCrossEntropyGrad topologically. So the loss output is FP16, then there is a CastFP32 applied to loss_scale before it times to the FP16 loss.

Comment (weixingzhang) [reply orttraining/orttraining/core/graph/mixed_precision_transformer.cc:60]: In current implementation, is SparseSoftmaxCrossEntropy followed by the op of loss scale? If so, then if loss scale is in FP32, the the CastFP16 can't be added to the output of SparseSoftmaxCrossEntropy. Otherwise, the two inputs of loss scale op will be different, which is not allowed for ONNX Mul op. Am I right? Or anything I missed?

Comment (iK1D) [reply orttraining/orttraining/core/graph/mixed_precision_transformer.cc:60]: For BERT, because there are two SparseSoftmaxCrossEntropy (mlm_loss and nsp_loss) and there is Add node to add them to get the final loss, so start from Add node, they are in FP16. If the loss is directly from the SparseSoftmaxCrossEntropy, then loss and loss_scale will remain FP32.

**Comment by xzhu1900** [comment]
Could we please try running mixed precision pipeline with this change? Unfortunately we don't have any CI test for pipeline yet (it should come soon with support for multi-GPU CI). 

For now, could you please try running the command below and see if it trains? (the command should be run on a multi-GPU machine. dev4 could do the job :-))

/bert_ort/openmpi/bin/mpirun --report-pid - -n 3 ./onnxruntime_training_bert --model_name /bert_ort/bert_models/nv/bert-large/bert-large-uncased_L_24_H_1024_A_16_V_30528_S_512_Dp_0.1_optimized_layer_norm --train_data_dir /bert_data/512/books_wiki_en_corpus/train --test_data_dir /bert_data/512/books_wiki_en_corpus/test --train_batch_size 1 --mode train --display_loss_steps 1 --optimizer lamb --learning_rate 0.006 --gradient_accumulation_steps 5 --num_train_steps 10 --warmup_ratio 0 --warmup_mode Linear --use_nccl --pipeline_parallel_size 3  --cut_group_info 1393:407-1463/1585/1707/1829/1951/2073/2195/2317,2369:407-2439/2561/2683/2805/2927/3049/3171/3293 --use_mixed_precision

File: orttraining/orttraining/core/graph/mixed_precision_transformer.cc [hunk 3]
```diff
@@ -70,10 +66,30 @@ bool IsFP32(const std::unordered_map<std::string, std::vector<int>>& map, std::s
   }
 }
 
+static const std::string Loss_Scale_Input = "loss_scale";
```
Comment (xzhu1900) [inline orttraining/orttraining/core/graph/mixed_precision_transformer.cc:69]: >Loss_Scale_Input [](start = 25, length = 16)

Nit: loss_scale_input?

(and same for loss_subgraph_entry_nodes and loss_subgraph_exit_nodes?)

**Comment by iK1D** [comment]
Add some training curves.

This is the BingAds BERT training curve. Zhijiang helped to train several steps. We can see loss curve is similar (even better) while loss scale value can go higher.
image

This is UniLM v2 training curve. Loss curve is nearly same, while loss scale can reach higher values.
image
